server/src/main/java/org/elasticsearch/repositories/SnapshotShardContext.java

```diff
@@ -66,7 +66,19 @@ public SnapshotShardContext(
         final long snapshotStartTime,
         ActionListener<ShardSnapshotResult> listener
     ) {
-        super(ActionListener.runBefore(listener, commitRef::onCompletion));
+        super(new ActionListener<>() {
+            @Override
+            public void onResponse(ShardSnapshotResult shardSnapshotResult) {
+                commitRef.onCompletion(listener.map(ignored -> shardSnapshotResult));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                commitRef.onCompletion(listener.map(ignored -> {
```

Comment:
I think we should also handle an exception from `onCompletion`, passed to the listener - ensuring that the original exception is passed to `listener` (with `onCompletion` exception as suppressed).

Reply:
Yes I see. Reworked things a bit so I could cover this in tests too, see ce4e836c85d383b86a0dd6a0657b7b2e7fb88ae3.